File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/create/RoomListActivity.java
package io.agora.scene.ktv.create;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;
import io.agora.scene.base.GlideApp;
import io.agora.scene.base.component.BaseViewBindingActivity;
import io.agora.scene.base.component.OnItemClickListener;
import io.agora.scene.ktv.R;
import io.agora.scene.ktv.databinding.KtvActivityRoomListBinding;
import io.agora.scene.ktv.live.RoomLivingActivity;
import io.agora.scene.ktv.service.KTVServiceProtocol;
import io.agora.scene.ktv.service.RoomListModel;
import io.agora.scene.widget.dialog.InputPasswordDialog;
import io.agora.scene.widget.utils.UiUtils;

/**
 * 房间列表
 */
public class RoomListActivity extends BaseViewBindingActivity<KtvActivityRoomListBinding> {
    private RoomListAdapter mAdapter;
    private RoomCreateViewModel roomCreateViewModel;
    private InputPasswordDialog inputPasswordDialog;
    private boolean isJoining = false;

    @Override
    protected KtvActivityRoomListBinding getViewBinding(@NonNull LayoutInflater inflater) {
        return KtvActivityRoomListBinding.inflate(inflater);
    }

    @Override
    protected void onResume() {
        super.onResume();
        setDarkStatusIcon(isBlackDarkStatus());
        loadRoomList();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        KTVServiceProtocol.Companion.getImplInstance().reset();
    }

    private void loadRoomList() {
        roomCreateViewModel.loadRooms();
    }

    @Override
    public void initView(@Nullable Bundle savedInstanceState) {
        roomCreateViewModel = new ViewModelProvider(this).get(RoomCreateViewModel.class);
        mAdapter = new RoomListAdapter(null, this, new OnItemClickListener() {
            @Override
            public void onItemClick(@NonNull Object data, View view, int position, long viewType) {
                RoomListModel model = (RoomListModel)data;
                if (model.isPrivate()) {
                    showInputPwdDialog(model);
                } else {
                    if (!isJoining) {
                        isJoining = true;
                        roomCreateViewModel.joinRoom(model.getRoomNo(), null);
                    }
                }
            }
        });
        getBinding().rvRooms.setLayoutManager(new GridLayoutManager(this, 2));
        getBinding().rvRooms.setAdapter(mAdapter);
        getBinding().smartRefreshLayout.setEnableLoadMore(false);
        setOnApplyWindowInsetsListener(getBinding().getRoot());
    }

    @Override
    public void initListener() {
        getBinding().btnCreateRoom.setOnClickListener(view -> {
            if (UiUtils.isFastClick(2000)) {
                return;
            }
            startActivity(new Intent(this, RoomCreateActivity.class));
        });
        roomCreateViewModel.roomModelList.observe(this, vlRoomListModels -> {
            hideLoadingView();
            getBinding().smartRefreshLayout.finishRefresh();
            if (vlRoomListModels == null || vlRoomListModels.isEmpty()) {
                getBinding().rvRooms.setVisibility(View.GONE);
                getBinding().tvTips1.setVisibility(View.VISIBLE);
                getBinding().ivBgMobile.setVisibility(View.VISIBLE);
            } else {
                mAdapter.setDataList(vlRoomListModels);
                getBinding().rvRooms.setVisibility(View.VISIBLE);
                getBinding().tvTips1.setVisibility(View.GONE);
                getBinding().ivBgMobile.setVisibility(View.GONE);
            }
        });
        roomCreateViewModel.joinRoomResult.observe(this, ktvJoinRoomOutputModel -> {
            isJoining = false;
            if (ktvJoinRoomOutputModel == null) {
                setDarkStatusIcon(isBlackDarkStatus());
            } else {
                RoomLivingActivity.launch(RoomListActivity.this, ktvJoinRoomOutputModel);
            }
        });
        getBinding().smartRefreshLayout.setOnRefreshListener(refreshLayout -> {
            loadRoomList();
        });
    }

    private void showInputPwdDialog(RoomListModel data) {
        if (inputPasswordDialog == null) {
            inputPasswordDialog = new InputPasswordDialog(this);
        }
        inputPasswordDialog.clearContent();
        inputPasswordDialog.iSingleCallback = (type, o) -> {
            roomCreateViewModel.joinRoom(data.getRoomNo(), (String) o);
        };
        inputPasswordDialog.show();
    }

    private class RoomListAdapter extends RecyclerView.Adapter<RoomListAdapter.ViewHolder> {
        private List<RoomListModel> mList;
        private Context mContext;
        private OnItemClickListener mListener;
        RoomListAdapter(List<RoomListModel> list, Context context, OnItemClickListener listener) {
            mList = list;
            mContext = context;
            mListener = listener;
        }

        public void setDataList(List<RoomListModel> list) {
            mList = list;
            notifyDataSetChanged();
        }
        @NonNull
        @Override
        public RoomListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(mContext).inflate(R.layout.ktv_item_room_list, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            Integer num = position % 5;
            int resId = getResources().getIdentifier("ktv_img_room_item_bg_" + num, "mipmap", getPackageName());
            holder.ivBackground.setImageResource(resId);
            RoomListModel data = mList.get(position);
            GlideApp.with(holder.ivAvatar.getContext()).load(data.getCreatorAvatar())
                    .into(holder.ivAvatar);
            holder.tvRoomName.setText(data.getName());
            holder.tvPersonNum.setText(holder.itemView.getContext().getString(R.string.ktv_people_count, data.getRoomPeopleNum()));
            holder.tvUserName.setText(data.getCreatorName());
            if (data.isPrivate()){
                holder.ivLock.setVisibility(View.VISIBLE);
            } else{
                holder.ivLock.setVisibility(View.GONE);
            }
            holder.itemView.setOnClickListener( view -> {
                mListener.onItemClick(data, view, position, getItemViewType(position));
            });
        }

        @Override
        public int getItemCount() {
            if (mList == null) {
                return 0;
            } else  {
                return mList.size();
            }
        }

        private class ViewHolder extends RecyclerView.ViewHolder {
            public ImageView ivBackground;
            public ImageView ivAvatar;
            public ImageView ivLock;
            public TextView tvRoomName;
            public TextView tvUserName;
            public TextView tvPersonNum;
            public ViewHolder(@NonNull View itemView) {
                super(itemView);
                ivBackground = itemView.findViewById(R.id.ivBackground);
                ivAvatar = itemView.findViewById(R.id.ivAvatar);
                ivLock = itemView.findViewById(R.id.ivLock);
                tvRoomName = itemView.findViewById(R.id.tvRoomName);
                tvUserName = itemView.findViewById(R.id.tvUserName);
                tvPersonNum = itemView.findViewById(R.id.tvPersonNum);
            }
        }
    }
}

File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/create/RoomListActivity.kt
package io.agora.scene.ktv.create

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.agora.rtmsyncmanager.model.AUIRoomInfo
import io.agora.scene.base.GlideApp
import io.agora.scene.base.component.BaseViewBindingActivity
import io.agora.scene.base.component.OnItemClickListener
import io.agora.scene.ktv.KtvCenter
import io.agora.scene.ktv.R
import io.agora.scene.ktv.databinding.KtvActivityRoomListBinding
import io.agora.scene.ktv.databinding.KtvItemRoomListBinding
import io.agora.scene.ktv.live.RoomLivingActivity
import io.agora.scene.ktv.service.KTVParameters
import io.agora.scene.ktv.service.KTVServiceProtocol.Companion.getImplInstance
import io.agora.scene.ktv.service.fullHeadUrl
import io.agora.scene.widget.dialog.InputPasswordDialog
import io.agora.scene.widget.utils.UiUtils

/**
 * Room list activity
 *
 * @constructor Create empty Room list activity
 */
class RoomListActivity : BaseViewBindingActivity<KtvActivityRoomListBinding>() {

    private val mRoomCreateViewModel: RoomCreateViewModel by lazy {
        ViewModelProvider(this)[RoomCreateViewModel::class.java]
    }

    private var mAdapter: RoomListAdapter? = null
    private var inputPasswordDialog: InputPasswordDialog? = null
    private var isJoining = false
    override fun getViewBinding(inflater: LayoutInflater): KtvActivityRoomListBinding {
        return KtvActivityRoomListBinding.inflate(inflater)
    }

    /**
     * On resume
     *
     */
    override fun onResume() {
        super.onResume()
        setDarkStatusIcon(isBlackDarkStatus)
    }

    /**
     * On restart
     *
     */
    override fun onRestart() {
        super.onRestart()
        binding.smartRefreshLayout.autoRefresh()
    }

    /**
     * On destroy
     *
     */
    override fun onDestroy() {
        super.onDestroy()
    }

    /**
     * Init view
     *
     * @param savedInstanceState
     */
    override fun initView(savedInstanceState: Bundle?) {
        mAdapter = RoomListAdapter(null, this, object : OnItemClickListener<AUIRoomInfo> {
            override fun onItemClick(data: AUIRoomInfo, view: View, position: Int, viewType: Long) {
                val password = data.customPayload[KTVParameters.PASSWORD] as? String
                if (!password.isNullOrEmpty()) {
                    showInputPwdDialog(data)
                } else {
                    if (!isJoining) {
                        isJoining = true
                        mRoomCreateViewModel.joinRoom(data, null)
                    }
                }
            }
        })
        binding.rvRooms.layoutManager = GridLayoutManager(this, 2)
        binding.rvRooms.adapter = mAdapter
        setOnApplyWindowInsetsListener(binding.root)
        binding.smartRefreshLayout.setEnableLoadMore(false)
        binding.smartRefreshLayout.setEnableRefresh(true)

        binding.smartRefreshLayout.setOnRefreshListener {
            mRoomCreateViewModel.loadRooms()
        }
        binding.smartRefreshLayout.autoRefresh()
    }

    /**
     * Init listener
     *
     */
    override fun initListener() {
        binding.btnCreateRoom.setOnClickListener { view: View? ->
            if (UiUtils.isFastClick(1000)) {
                return@setOnClickListener
            }
            CreateRoomDialog(this).show(supportFragmentManager, "CreateRoomDialog")
        }
        mRoomCreateViewModel.roomModelList.observe(this) { vlRoomListModels: List<AUIRoomInfo>? ->
            hideLoadingView()
            binding!!.smartRefreshLayout.finishRefresh()
            if (vlRoomListModels.isNullOrEmpty()) {
                binding.rvRooms.visibility = View.GONE
                binding.tvTips1.visibility = View.VISIBLE
                binding.ivBgMobile.setVisibility(View.VISIBLE)
            } else {
                mAdapter?.setDataList(vlRoomListModels)
                binding.rvRooms.visibility = View.VISIBLE
                binding.tvTips1.visibility = View.GONE
                binding.ivBgMobile.setVisibility(View.GONE)
            }
        }
        mRoomCreateViewModel.roomInfoLiveData.observe(this) { roomInfo ->
            isJoining = false
            if (roomInfo == null) {
                setDarkStatusIcon(isBlackDarkStatus)
            } else {
                RoomLivingActivity.launch(this, roomInfo)
            }
        }
    }

    private fun showInputPwdDialog(data: AUIRoomInfo) {
        if (inputPasswordDialog == null) {
            inputPasswordDialog = InputPasswordDialog(this)
        }
        inputPasswordDialog?.apply {
            clearContent()
            onDefineClickListener = InputPasswordDialog.OnDefineClickListener { password ->
                mRoomCreateViewModel.joinRoom(data, password)
            }
            show()
        }
    }

    private inner class RoomListAdapter constructor(
        private var mList: List<AUIRoomInfo>?,
        private val mContext: Context,
        private val mListener: OnItemClickListener<AUIRoomInfo>
    ) : RecyclerView.Adapter<RoomListAdapter.ViewHolder>() {
        /**
         * Set data list
         *
         * @param list
         */
        fun setDataList(list: List<AUIRoomInfo>?) {
            mList = list
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(KtvItemRoomListBinding.inflate(LayoutInflater.from(parent.context), parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val num = position % 5
            val resId = getResources().getIdentifier("ktv_img_room_item_bg_$num", "mipmap", packageName)
            holder.binding.ivBackground.setImageResource(resId)
            val data = mList?.get(position) ?: return
            if (data.roomOwner != null) {
                GlideApp.with(mContext)
                    .load(data.roomOwner!!.fullHeadUrl)
                    .into(holder.binding.ivAvatar)
            } else {
                holder.binding.ivAvatar.setImageResource(R.mipmap.default_user_avatar)
            }
            holder.binding.tvRoomName.text = data.roomName
            val userCount = data.customPayload[KTVParameters.ROOM_USER_COUNT] as? Long
            val showCount = (userCount ?: 0) + KtvCenter.userAddMore
            holder.binding.tvPersonNum.text = mContext.getString(R.string.ktv_people_count, showCount)
            holder.binding.tvUserName.text = data.roomOwner?.userName ?: ""
            val password = data.customPayload[KTVParameters.PASSWORD] as? String
            holder.binding.ivLock.isVisible = !password.isNullOrEmpty()
            holder.itemView.setOnClickListener { view: View? ->
                mListener.onItemClick(
                    data,
                    view,
                    position,
                    getItemViewType(position).toLong()
                )
            }
        }

        override fun getItemCount(): Int = mList?.size ?: 0


        /**
         * View holder
         *
         * @property binding
         * @constructor Create empty View holder
         */
        inner class ViewHolder constructor(val binding: KtvItemRoomListBinding) : RecyclerView.ViewHolder(binding.root)
    }
}
